src/CameraOnvif.cpp

```diff
@@ -0,0 +1,408 @@
+    _timg__GetImagingSettingsResponse GetImagingSettingsResponse;
+    setCredentials();
+    if (m_private->proxy_image->GetImagingSettings(&GetImagingSettings, GetImagingSettingsResponse))
+        reportError();
+    return *GetImagingSettingsResponse.ImagingSettings->Brightness;
+}
+float CameraOnvif::readColorSaturation(){
+    _timg__GetImagingSettings GetImagingSettings;
+    GetImagingSettings.VideoSourceToken = video_token;
+    _timg__GetImagingSettingsResponse GetImagingSettingsResponse;
+    setCredentials();
+    if (m_private->proxy_image->GetImagingSettings(&GetImagingSettings, GetImagingSettingsResponse))
+        reportError();
+    return *GetImagingSettingsResponse.ImagingSettings->ColorSaturation;
+}
+float CameraOnvif::readContrast(){
+    _timg__GetImagingSettings GetImagingSettings;
+    GetImagingSettings.VideoSourceToken = video_token;
+    _timg__GetImagingSettingsResponse GetImagingSettingsResponse;
+    setCredentials();
+    if (m_private->proxy_image->GetImagingSettings(&GetImagingSettings, GetImagingSettingsResponse))
+        reportError();
+    return *GetImagingSettingsResponse.ImagingSettings->Contrast;
+}
+
+void CameraOnvif::setBrightness(float percentage){
+    float new_brightness = (brightness[1] - brightness[0]) * percentage + brightness[0];
+    _timg__SetImagingSettings SetImagingSettings;
+    SetImagingSettings.VideoSourceToken = video_token;
+    SetImagingSettings.ImagingSettings = new tt__ImagingSettings20();
+    SetImagingSettings.ImagingSettings->Brightness = &new_brightness;
+    _timg__SetImagingSettingsResponse SetImagingSettingsResponse;
+    setCredentials();
+    if (m_private->proxy_image->SetImagingSettings(&SetImagingSettings, SetImagingSettingsResponse))
+        reportError();
+}
+
+void CameraOnvif::setColorSaturation(float percentage){
+    float new_color_saturation = (color_saturation[1] - color_saturation[0]) * percentage + color_saturation[0];
+    _timg__SetImagingSettings SetImagingSettings;
+    SetImagingSettings.VideoSourceToken = video_token;
+    SetImagingSettings.ImagingSettings = new tt__ImagingSettings20();
+    SetImagingSettings.ImagingSettings->ColorSaturation = &new_color_saturation;
+    _timg__SetImagingSettingsResponse SetImagingSettingsResponse;
+    setCredentials();
+    if (m_private->proxy_image->SetImagingSettings(&SetImagingSettings, SetImagingSettingsResponse))
+        reportError();
+}
+
+void CameraOnvif::setContrast(float percentage){
+    float new_contrast = (contrast[1] - contrast[0]) * percentage + contrast[0];
+    _timg__SetImagingSettings SetImagingSettings;
+    SetImagingSettings.VideoSourceToken = video_token;
+    SetImagingSettings.ImagingSettings = new tt__ImagingSettings20();
+    SetImagingSettings.ImagingSettings->Contrast = &new_contrast;
+    _timg__SetImagingSettingsResponse SetImagingSettingsResponse;
+    setCredentials();
+    if (m_private->proxy_image->SetImagingSettings(&SetImagingSettings, SetImagingSettingsResponse))
```

Comment:
Can't this also change the color saturation because the value is uninitialized ?

It might make more sense to have a single method to set all parameters set by setImageSettings. This would allow the caller to set them both in one call (more efficient) & ensure that all parameters that need to be set are. What do you think ?

Reply:
This would apply to the setters above as well. I'd think of a ImageSettings structure with the relevant parameters, and then only getImageSettings/setImageSettings

Reply:
To the first one, I will verify each one but as this parameters are optional and previous test didn't showed any error and also ONVIF return error when I try to set something wrong.

I thought about that, making one function to set everything. But and if you want to set just one? I made that to don't have to getImageSettings before setting. Note that although it's inefficient creating the setter the communication just happen once for each. So, if changing 3 just 3 communication. Setting 3 in one function having to GetSettings I would have 2 communications always. So, the decision of changing it or not relies on what is more common to happen? Changing 1 or 3 parameters?

(Note: I also think that getting information are more costly in general, like sending request, getting info from cam, transporting back and passing to response)

Reply:
I meant, set all three each time. If the caller wants to allow for setting only one he/she can query them one time and then track the three others.